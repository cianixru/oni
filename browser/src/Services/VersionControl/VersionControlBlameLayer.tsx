import { pathExists } from "fs-extra"
import { Buffer, BufferLayer, Commands, Configuration } from "oni-api"
import { warn } from "oni-core-logging"
import * as React from "react"
import { Transition } from "react-transition-group"
import { Position } from "vscode-languageserver-types"

import { LayerContextWithCursor } from "../../Editor/NeovimEditor/NeovimBufferLayersView"
import styled, { pixel, textOverflow, withProps } from "../../UI/components/common"
import { getTimeSince } from "../../Utility"
import { VersionControlProvider } from "./"
import { Blame as IBlame } from "./VersionControlProvider"

type TransitionStates = "entering" | "entered" | "exiting"

interface IBlamePosition {
    top: number
    left: number
    hide: boolean
}

interface ICanFit {
    canFit: boolean
    message: string
    position: IBlamePosition
}

interface ILineDetails {
    nextSpacing: number
    lastEmptyLine: number
}

export interface IProps extends LayerContextWithCursor {
    getBlame: (lineOne: number, lineTwo: number) => Promise<IBlame>
    timeout: number
    cursorScreenLine: number
    cursorBufferLine: number
    currentLine: string
    mode: "auto" | "manual"
    fontFamily: string
    setupCommand: (callback: () => void) => void
}

export interface IState {
    blame: IBlame
    showBlame: boolean
    currentLineContent: string
    currentCursorBufferLine: number
    error: Error
}

interface IContainerProps {
    height: number
    top: number
    left: number
    fontFamily: string
    hide: boolean
    timeout: number
    animationState: TransitionStates
}

const getOpacity = (state: TransitionStates) => {
    const transitionStyles = {
        entering: 0,
        entered: 0.5,
        exiting: 0,
    }
    return transitionStyles[state]
}

export const BlameContainer = withProps<IContainerProps>(styled.div).attrs({
    style: ({ top, left }: IContainerProps) => ({
        top: pixel(top),
        left: pixel(left),
    }),
})`
    ${p => p.hide && `visibility: hidden`};
    width: auto;
    box-sizing: border-box;
    position: absolute;
    font-style: italic;
    font-family: ${p => p.fontFamily};
    color: ${p => p.theme["menu.foreground"]};
    opacity: ${p => getOpacity(p.animationState)};
    transition: opacity ${p => p.timeout}ms ease-in-out;
    height: ${p => pixel(p.height)};
    line-height: ${p => pixel(p.height)};
    right: 3em;
    ${textOverflow}
`

const BlameDetails = styled.span`
    color: inherit;
    width: 100%;
`

// CurrentLine - the string in the current line
// CursorLine - The 0 based position of the cursor in the file i.e. at line 30 this will be 29
// CursorBufferLine - The 1 based position of the cursor in the file i.e. at line 30 it will be 30
// CursorScreenLine - the position of the cursor within the visible lines so if line 30 is at the
// top of the viewport it will be 0

export class Blame extends React.PureComponent<IProps, IState> {
    // Reset show blame to false when props change - do it here so it happens before rendering
    // hide if the current line has changed or if the text of the line has changed
    // aka input is in progress or if there is an empty line
    public static getDerivedStateFromProps(nextProps: IProps, prevState: IState) {
        const lineNumberChanged = nextProps.cursorBufferLine !== prevState.currentCursorBufferLine
        const lineContentChanged = prevState.currentLineContent !== nextProps.currentLine
        if (
            (prevState.showBlame && (lineNumberChanged || lineContentChanged)) ||
            !nextProps.currentLine
        ) {
            return {
                showBlame: false,
                blame: prevState.blame,
                currentLineContent: nextProps.currentLine,
                currentCursorBufferLine: nextProps.cursorBufferLine,
            }
        }
        return null
    }

    public state: IState = {
        error: null,
        blame: null,
        showBlame: null,
        currentLineContent: this.props.currentLine,
        currentCursorBufferLine: this.props.cursorBufferLine,
    }

    private _timeout: any
    private readonly DURATION = 300
    private readonly LEFT_OFFSET = 4

    public async componentDidMount() {
        const { cursorBufferLine, mode } = this.props
        await this.updateBlame(cursorBufferLine, cursorBufferLine)
        if (mode === "auto") {
            this.resetTimer()
        }
        this.props.setupCommand(() => {
            const { showBlame } = this.state
            this.setState({ showBlame: !showBlame })
        })
    }

    public async componentDidUpdate(prevProps: IProps, prevState: IState) {
        const { cursorBufferLine, currentLine, mode } = this.props
        if (prevProps.cursorBufferLine !== cursorBufferLine && currentLine) {
            await this.updateBlame(cursorBufferLine, cursorBufferLine)
            if (mode === "auto") {
                return this.resetTimer()
            }
        }
    }

    public componentWillUnmount() {
        clearTimeout(this._timeout)
    }

    public componentDidCatch(error: Error) {
        warn(`Oni VCS Blame layer failed because: ${error.message}`)
        this.setState({ error })
    }

    public resetTimer = () => {
        clearTimeout(this._timeout)
        this._timeout = setTimeout(() => {
            if (this.props.currentLine) {
                this.setState({ showBlame: true })
            }
        }, this.props.timeout)
    }

    public getLastEmptyLine() {
        const { cursorLine, visibleLines, topBufferLine } = this.props
        const lineDetails: ILineDetails = {
            lastEmptyLine: null,
            nextSpacing: null,
        }
        for (
            let currentBufferLine = cursorLine;
            currentBufferLine >= topBufferLine;
            currentBufferLine--
        ) {
            const screenLine = currentBufferLine - topBufferLine
            const line = visibleLines[screenLine]
            if (!line.length) {
                const nextLine = visibleLines[screenLine + 1]
                lineDetails.lastEmptyLine = currentBufferLine
                // search for index of first non-whitespace character which is equivalent
                // to the whitespace count
                lineDetails.nextSpacing = nextLine.search(/\S/)
                break
            }
        }
        return lineDetails
    }

    public calculatePosition(canFit: boolean) {
        const { cursorLine, cursorScreenLine, visibleLines } = this.props
        const currentLine = visibleLines[cursorScreenLine]
        const character = currentLine && currentLine.length + this.LEFT_OFFSET

        if (canFit) {
            return this.getPosition({ line: cursorLine, character })
        }

        const { lastEmptyLine, nextSpacing } = this.getLastEmptyLine()

        if (lastEmptyLine) {
            return this.getPosition({ line: lastEmptyLine - 1, character: nextSpacing })
        }

        return this.getPosition()
    }

    // TODO: possibly add a caching strategy so a new call isn't made each time or
    // get a blame for the entire file and store it
    public updateBlame = async (lineOne: number, lineTwo: number) => {
        const outOfBounds = this.isOutOfBounds(lineOne, lineTwo)
        const blame = !outOfBounds ? await this.props.getBlame(lineOne, lineTwo) : null
        this.setState({ blame })
    }

    public formatCommitDate(timestamp: string) {
        return new Date(parseInt(timestamp, 10) * 1000)
    }

    public getPosition(positionToRender?: Position): IBlamePosition {
        const emptyPosition: IBlamePosition = {
            hide: true,
            top: null,
            left: null,
        }
        if (!positionToRender) {
            return emptyPosition
        }
        const position = this.props.bufferToPixel(positionToRender)
        if (!position) {
            return emptyPosition
        }
        return {
            hide: false,
            top: position.pixelY,
            left: position.pixelX,
        }
    }

    public isOutOfBounds = (...lines: number[]) => {
        return lines.some(
            line => !line || line > this.props.bottomBufferLine || line < this.props.topBufferLine,
        )
    }

    public getBlameText = (numberOfTruncations = 0) => {
        const { blame } = this.state
        if (!blame) {
            return null
        }
        const { author, hash, committer_time } = blame
        const formattedDate = this.formatCommitDate(committer_time)
        const timeSince = `${getTimeSince(formattedDate)} ago`
        const formattedHash = hash.slice(0, 4).toUpperCase()

        const words = blame.summary.split(" ")
        const message = words.slice(0, words.length - numberOfTruncations).join(" ")

        const symbol = "â€¦"
        const summary = numberOfTruncations && words.length > 2 ? message.concat(symbol) : message

        return words.length < 2
            ? `${author}, ${timeSince}`
            : `${author}, ${timeSince}, ${summary} #${formattedHash}`
    }

    // Recursively calls get blame text if the message will not fit onto the screen up
    // to a limit of 6 times each time removing one word from the blame message
    // if after 6 attempts the message is still not small enougth then we render the popup
    public canFit = (truncationAmount = 0): ICanFit => {
        const { visibleLines, dimensions, cursorScreenLine } = this.props
        const message = this.getBlameText(truncationAmount)
        const currentLine = visibleLines[cursorScreenLine] || ""
        const canFit = dimensions.width > currentLine.length + message.length + this.LEFT_OFFSET

        if (!canFit && truncationAmount <= 6) {
            return this.canFit(truncationAmount + 1)
        }
        const truncatedOrFullMessage = canFit ? message : this.getBlameText()
        return {
            canFit,
            message: truncatedOrFullMessage,
            position: this.calculatePosition(canFit),
        }
    }

    public render() {
        const { blame, showBlame, error } = this.state
        if (!blame || !showBlame || error) {
            return null
        }
        const { message, position } = this.canFit()
        return (
            <Transition in={blame && showBlame} timeout={this.DURATION}>
                {(state: TransitionStates) => (
                    <BlameContainer
                        {...position}
                        data-id="vcs.blame"
                        timeout={this.DURATION}
                        animationState={state}
                        height={this.props.fontPixelHeight}
                        fontFamily={this.props.fontFamily}
                    >
                        <BlameDetails>{message}</BlameDetails>
                    </BlameContainer>
                )}
            </Transition>
        )
    }
}

export default class VersionControlBlameLayer implements BufferLayer {
    constructor(
        private _buffer: Buffer,
        private _vcsProvider: VersionControlProvider,
        private _configuration: Configuration,
        private _commands: Commands.Api,
    ) {}

    public getBlame = async (lineOne: number, lineTwo: number) => {
        const fileExists = await pathExists(this._buffer.filePath)
        return (
            fileExists &&
            this._vcsProvider.getBlame({ file: this._buffer.filePath, lineOne, lineTwo })
        )
    }

    get id() {
        return "vcs.blame"
    }

    public setupCommand = (callback: () => void) => {
        this._commands.registerCommand({
            command: "experimental.vcs.blame.toggleBlame",
            name: null,
            detail: null,
            enabled: this._isActive,
            execute: callback,
        })
    }

    public getConfigOpts() {
        const fontFamily = this._configuration.getValue<string>("editor.fontFamily")
        const timeout = this._configuration.getValue<number>("experimental.vcs.blame.timeout")
        const mode = this._configuration.getValue<"auto" | "manual">("experimental.vcs.blame.mode")

        return { timeout, mode, fontFamily }
    }

    public render(context: LayerContextWithCursor) {
        const cursorBufferLine = context.cursorLine + 1
        const cursorScreenLine = cursorBufferLine - context.topBufferLine
        const config = this.getConfigOpts()
        const activated = this._isActive()
        return (
            activated && (
                <Blame
                    {...context}
                    mode={config.mode}
                    timeout={config.timeout}
                    getBlame={this.getBlame}
                    fontFamily={config.fontFamily}
                    setupCommand={this.setupCommand}
                    cursorBufferLine={cursorBufferLine}
                    cursorScreenLine={cursorScreenLine}
                    currentLine={context.visibleLines[cursorScreenLine]}
                />
            )
        )
    }

    private _isActive() {
        return this._vcsProvider && this._vcsProvider.isActivated
    }
}
