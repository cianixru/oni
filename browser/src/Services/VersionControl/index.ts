export {
    Diff,
    Summary,
    SupportedProviders,
    default as VersionControlProvider,
} from "./VersionControlProvider"
export { activate, getInstance, VersionControlManager } from "./VersionControlManager"
export { default as VersionControlPane } from "./VersionControlPane"
export { default as store } from "./VersionControlStore"
export { default as VersionControlView } from "./VersionControlView"
