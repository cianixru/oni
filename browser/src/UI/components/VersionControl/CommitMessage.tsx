import * as React from "react"

import styled from "./../../components/common"
import TextInput from "./../LightweightText"

export const Explainer = styled.span`
    width: 100%;
    padding-left: 0.5rem;
    text-align: left;
    font-size: 0.8em;
    display: block;
    opacity: 0.4;
`

const TextArea = styled.textarea`
    width: 100%;
    background-color: inherit;
    color: inherit;
    font-size: inherit;
    font-family: inherit;
    padding: 0.5em;
    box-sizing: border-box;
    overflow: hidden;
    resize: vertical;
`
export interface ICommitHandlers {
    handleCommitMessage: (evt: React.ChangeEvent<HTMLInputElement>) => void
    handleCommitCancel: () => void
    handleCommitComplete: () => void
}

const CommitMessage: React.SFC<ICommitHandlers> = props => (
    <>
        <Explainer>Hit enter to commit the file</Explainer>
        <TextInput
            InputComponent={TextArea}
            onChange={props.handleCommitMessage}
            onCancel={props.handleCommitCancel}
            onComplete={props.handleCommitComplete}
            defaultValue="Enter a commit message"
        />
    </>
)
export default CommitMessage
