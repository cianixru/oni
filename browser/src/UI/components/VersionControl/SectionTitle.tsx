import * as React from "react"

import Caret from "./../Caret"
import { sidebarItemSelected, styled, withProps } from "./../common"

export const Title = styled.h4`
    margin: 0;
`

export const SectionTitle = withProps<Partial<IProps>>(styled.div)`
    ${sidebarItemSelected};
    margin: 0.2em 0;
    padding: 0.2em;
    background-color: rgba(0, 0, 0, 0.2);
    display: flex;
    justify-content: space-between;
`

interface IProps {
    active: boolean
    isSelected: boolean
    title: string
    onClick: () => void
    count?: number
    testId: string
}

const VCSSectionTitle: React.SFC<IProps> = props => (
    <SectionTitle isSelected={props.isSelected} data-test={props.testId} onClick={props.onClick}>
        <Caret active={props.active} />
        <Title>{props.title.toUpperCase()}</Title>
        <strong>{props.count}</strong>
    </SectionTitle>
)

export default VCSSectionTitle
