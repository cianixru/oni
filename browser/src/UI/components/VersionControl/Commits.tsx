import * as React from "react"

import { Logs } from "../../../Services/VersionControl/VersionControlProvider"
import { sidebarItemSelected, styled, withProps } from "./../../../UI/components/common"
import { formatDate } from "./../../../Utility"
import VCSSectionTitle from "./../SectionTitle"

interface ICommitsSection {
    commits: Logs["all"]
    selectedId: string
    titleId: string
    visibility: boolean
    toggleVisibility: () => void
    onClick: (c: string) => void
}

const List = styled.ul`
    list-style: none;
    margin: 0;
    padding: 0;
    overflow-y: auto;
    overflow-x: hidden;
    max-height: 30em;
`
export const ListItem = withProps<{ isSelected?: boolean }>(styled.li)`
    ${({ isSelected }) => isSelected && sidebarItemSelected};
    padding: 0.4rem;
`

const Detail = styled.p`
    margin: 0.4rem 0;
`

const Container = styled.div`
    width: 100%;
    max-height: 20em;
    overflow: hidden;
`

const CommitsSection: React.SFC<ICommitsSection> = ({ commits, ...props }) => {
    return (
        <Container>
            <VCSSectionTitle
                isSelected={props.selectedId === props.titleId}
                testId={`${props.titleId}-${commits.length}`}
                onClick={props.toggleVisibility}
                active={!!(props.visibility && commits.length)}
                title="Recent Commits"
                count={commits.length}
            />
            {props.visibility && commits.length ? (
                <List>
                    {commits.map(prevCommit => (
                        <ListItem
                            key={prevCommit.hash}
                            onClick={() => props.onClick(prevCommit.hash)}
                            isSelected={props.selectedId === prevCommit.hash}
                        >
                            <Detail>
                                <strong> {prevCommit.message}</strong>
                            </Detail>
                            <Detail>{prevCommit.hash.slice(0, 6)}</Detail>
                            <Detail>{formatDate(prevCommit.date)}</Detail>
                            <Detail>{prevCommit.author_email}</Detail>
                            <Detail>{prevCommit.author_name}</Detail>
                        </ListItem>
                    ))}
                </List>
            ) : null}
        </Container>
    )
}

export default CommitsSection
