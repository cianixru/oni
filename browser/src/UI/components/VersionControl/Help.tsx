import * as React from "react"

import { inputManager } from "../../../Services/InputManager"
import styled from "../common"
import { SectionTitle, Title } from "./../SectionTitle"

const sidebarCommands = [
    { command: "vcs.openFile", description: "Open the currently selected file" },
    { command: "vcs.unstage", description: "Unstage the currently selected file" },
    { command: "vcs.commitAll", description: "Commit all staged files" },
    { command: "vcs.refresh", description: "Manually refresh vcs pane" },
    { command: "vcs.sidebar.toggle", description: "Toggle vcs pane" },
]
const getBoundKeys = (commands = sidebarCommands) => {
    return commands.map(({ command, ...rest }) => ({
        key: inputManager.getBoundKeys(command)[0] || "Unbound",
        ...rest,
    }))
}

const HelpContainer = styled.div``

const CommandExplainer = styled.div`
    padding: 0 0.5em;
`

export const Description = styled.p`
    margin: 0.5em 0;
`

const Command = styled.span`
    color: ${p => p.theme["highlight.mode.insert.foreground"]};
    background-color: ${p => p.theme["highlight.mode.insert.background"]};
`

const wrapInBrackets = (str: string) => {
    return str.startsWith("<") && str.endsWith(">") ? str : `<${str}>`
}

const Help: React.SFC<{}> = props => {
    const commands = getBoundKeys()
    return (
        <HelpContainer>
            <SectionTitle>
                <Title>Help</Title>?
            </SectionTitle>
            {commands.map(command => (
                <CommandExplainer key={command.key}>
                    <Description>
                        {command.description} - <Command>{wrapInBrackets(command.key)}</Command>
                    </Description>
                </CommandExplainer>
            ))}
        </HelpContainer>
    )
}

export default Help
