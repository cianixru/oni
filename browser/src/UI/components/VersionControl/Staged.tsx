import * as React from "react"

import styled, { Center, getSelectedBorder, withProps } from "../common"
import SectionTitle from "../SectionTitle"
import { LoadingSpinner } from "./../../../UI/components/LoadingSpinner"
import { Icons } from "./../../../UI/components/Octicon"
import CommitMessage from "./CommitMessage"
import File from "./File"

const Explainer = styled.div`
    width: 100%;
    text-align: left;
    display: block;
    opacity: 0.4;
    padding: 0.5rem;
`

interface IProps {
    files?: string[]
    titleId: string
    selectedId: string
    icon: Icons
    loading: boolean
    handleSelection: (id: string) => void
    filesToCommit: string[]
    toggleVisibility: () => void
    handleCommitMessage: (evt: React.ChangeEvent<HTMLInputElement>) => void
    handleCommitOne: () => void
    handleCommitCancel: () => void
    handleCommitAll: () => void
    selectedToCommit?: (id: string) => boolean
    visible: boolean
}

const OptionsBar = withProps<{ isSelected: boolean }>(styled.span)`
    border: ${getSelectedBorder};
    display: block;
    width: 100%;
`

export const LoadingHandler: React.SFC<{ loading: boolean }> = ({ loading, children }) => (
    <>
        {loading ? (
            <Center>
                <LoadingSpinner iconSize="0.4em" />
            </Center>
        ) : (
            children
        )}
    </>
)

const StagedSection: React.SFC<IProps> = props => {
    const commitAllSelected = "commit_all" === props.selectedId
    return (
        <div>
            <SectionTitle
                isSelected={props.selectedId === props.titleId}
                testId={`${props.titleId}-${props.files.length}`}
                onClick={props.toggleVisibility}
                active={props.visible && !!props.files.length}
                title={props.titleId}
                count={props.files.length}
            />
            {props.visible && props.files.length ? (
                <LoadingHandler loading={props.loading && commitAllSelected}>
                    <OptionsBar isSelected={commitAllSelected}>
                        {props.selectedToCommit("commit_all") ? (
                            <CommitMessage
                                handleCommitCancel={props.handleCommitCancel}
                                handleCommitComplete={props.handleCommitAll}
                                handleCommitMessage={props.handleCommitMessage}
                            />
                        ) : (
                            <Explainer onClick={() => props.handleSelection(props.titleId)}>
                                commit all ({props.files.length})
                            </Explainer>
                        )}
                    </OptionsBar>
                </LoadingHandler>
            ) : null}
            {props.visible &&
                props.files.map(file => {
                    const isSelected = file === props.selectedId
                    const isLoading = props.filesToCommit.includes(file) && props.loading
                    return (
                        <LoadingHandler loading={isLoading} key={file}>
                            {props.selectedToCommit(file) ? (
                                <CommitMessage
                                    handleCommitCancel={props.handleCommitCancel}
                                    handleCommitComplete={props.handleCommitOne}
                                    handleCommitMessage={props.handleCommitMessage}
                                />
                            ) : (
                                <File
                                    file={file}
                                    icon={props.icon}
                                    onClick={props.handleSelection}
                                    isSelected={isSelected}
                                />
                            )}
                        </LoadingHandler>
                    )
                })}
        </div>
    )
}

export default StagedSection
