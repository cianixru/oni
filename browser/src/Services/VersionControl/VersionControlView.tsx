import * as React from "react"
import { connect } from "react-redux"

import { styled } from "./../../UI/components/common"
import { SectionTitle, Title } from "./../../UI/components/SectionTitle"
import CommitsSection from "./../../UI/components/VersionControl/Commits"
import Help from "./../../UI/components/VersionControl/Help"
import StagedSection from "./../../UI/components/VersionControl/Staged"
import VersionControlStatus from "./../../UI/components/VersionControl/Status"
import { VimNavigator } from "./../../UI/components/VimNavigator"
import { IDsMap } from "./VersionControlPane"
import { Logs, StatusResult } from "./VersionControlProvider"
import {
    PrevCommits,
    ProviderActions,
    VersionControlActions,
    VersionControlState,
} from "./VersionControlStore"

const StatusContainer = styled.div`
    overflow-x: hidden;
    overflow-y: auto;
`

interface IStateProps {
    loading: boolean
    filesToCommit: string[]
    loadingSection: ProviderActions
    status: StatusResult
    hasFocus: boolean
    hasError: boolean
    activated: boolean
    committing: boolean
    message: string[]
    selectedItem: string
    commits: PrevCommits[]
    showHelp: boolean
    logs: Logs
}

interface IDispatchProps {
    cancelCommit: () => void
    updateCommitMessage: (message: string[]) => void
    setLoading: (loading: boolean) => void
}

interface IProps {
    IDs: IDsMap
    setError?: (e: Error) => void
    commit?: (message: string[], files?: string[]) => Promise<void>
    updateSelection?: (selection: string) => void
    handleSelection?: (selection: string) => void
    getStatus: () => void
}

type ConnectedProps = IProps & IStateProps & IDispatchProps

interface State {
    modified: boolean
    staged: boolean
    untracked: boolean
    commits: boolean
}

export class VersionControlView extends React.Component<ConnectedProps, State> {
    public state: State = {
        modified: true,
        staged: true,
        untracked: true,
        commits: true,
    }

    public async componentDidMount() {
        await this.props.getStatus()
    }

    public async componentDidCatch(e: Error) {
        this.props.setError(e)
    }

    public toggleVisibility = (section: keyof State) => {
        this.setState(prevState => ({ ...prevState, [section]: !prevState[section] }))
    }

    public toggleOrAction = (id: string) => {
        const isSectionId = Object.keys(this.state).includes(id)
        if (isSectionId) {
            this.toggleVisibility(id as keyof State)
        }
        this.props.handleSelection(id)
    }

    public formatCommit = (message: string) => {
        return message.length >= 50 ? [message.substr(0, 50), message.substr(50)] : [message]
    }

    public handleCommitMessage = (evt: React.ChangeEvent<HTMLInputElement>) => {
        const { value } = evt.currentTarget
        const message = this.formatCommit(value)
        this.props.updateCommitMessage(message)
    }

    public handleCommitOne = async () => {
        const { message, selectedItem } = this.props
        await this.props.commit(message, [selectedItem])
    }

    public handleCommitAll = async () => {
        const { message } = this.props
        await this.props.commit(message)
    }

    public handleCommitCancel = () => {
        this.props.cancelCommit()
    }

    public insertIf(condition: boolean, element: string[]) {
        return condition ? element : []
    }

    public isSelected = (id: string) =>
        this.props.committing &&
        this.props.status.staged.length &&
        this.state.staged &&
        this.props.selectedItem === id

    public getIds = () => {
        const { logs, status, IDs } = this.props
        const { modified, staged, untracked } = status
        const commitSHAs = logs.all.slice(0, 1).map(({ hash }) => hash)
        const ids = [
            IDs.commits,
            ...this.insertIf(this.state.commits, commitSHAs),
            IDs.staged,
            ...this.insertIf(this.state.staged && !!staged.length, [IDs.commitAll]),
            ...this.insertIf(this.state.staged, staged),
            IDs.modified,
            ...this.insertIf(this.state.modified, modified),
            IDs.untracked,
            ...this.insertIf(this.state.untracked, untracked),
        ]
        return ids
    }

    public render() {
        const error = this.props.hasError && "Something Went Wrong!"
        const inactive = !this.props.activated && "Version Control Not Available"
        const warning = error || inactive
        const {
            IDs,
            logs,
            showHelp,
            loading,
            committing,
            filesToCommit,
            loadingSection,
            status: { modified, staged, untracked },
        } = this.props

        const commitInProgress = loading && loadingSection === "commit"

        return warning ? (
            <SectionTitle>
                <Title>{warning}</Title>
            </SectionTitle>
        ) : (
            <>
                {!showHelp ? <Title>To show help press "?"</Title> : <Help />}
                <VimNavigator
                    ids={this.getIds()}
                    active={this.props.hasFocus && !committing}
                    onSelected={this.toggleOrAction}
                    onSelectionChanged={this.props.updateSelection}
                    render={selectedId => (
                        <StatusContainer>
                            <CommitsSection
                                titleId={IDs.commits}
                                selectedId={selectedId}
                                visibility={this.state.commits}
                                onClick={this.props.handleSelection}
                                commits={logs.all}
                                toggleVisibility={() => this.toggleVisibility(IDs.commits)}
                            />
                            <StagedSection
                                titleId={IDs.staged}
                                icon="diff-added"
                                files={staged}
                                selectedId={selectedId}
                                filesToCommit={filesToCommit}
                                selectedToCommit={this.isSelected}
                                visible={this.state.staged}
                                loading={commitInProgress}
                                handleSelection={this.props.handleSelection}
                                toggleVisibility={() => this.toggleVisibility(IDs.staged)}
                                handleCommitOne={this.handleCommitOne}
                                handleCommitAll={this.handleCommitAll}
                                handleCommitMessage={this.handleCommitMessage}
                                handleCommitCancel={this.handleCommitCancel}
                            />
                            <VersionControlStatus
                                icon="diff-modified"
                                files={modified}
                                titleId={IDs.modified}
                                selectedId={selectedId}
                                visibility={this.state.modified}
                                onClick={this.props.handleSelection}
                                toggleVisibility={() => this.toggleVisibility(IDs.modified)}
                            />
                            <VersionControlStatus
                                files={untracked}
                                icon="diff-ignored"
                                titleId={IDs.untracked}
                                selectedId={selectedId}
                                visibility={this.state.untracked}
                                onClick={this.props.handleSelection}
                                toggleVisibility={() => this.toggleVisibility(IDs.untracked)}
                            />
                        </StatusContainer>
                    )}
                />
            </>
        )
    }
}

const mapStateToProps = (state: VersionControlState): IStateProps => ({
    status: state.status,
    hasFocus: state.hasFocus,
    hasError: state.hasError,
    activated: state.activated,
    committing: state.commit.active,
    message: state.commit.message,
    selectedItem: state.selected,
    commits: state.commit.previousCommits,
    filesToCommit: state.commit.files,
    showHelp: state.help.active,
    loading: state.loading.active,
    loadingSection: state.loading.type,
    logs: state.logs,
})

const ConnectedGitComponent = connect<IStateProps, IDispatchProps, IProps>(
    mapStateToProps,
    VersionControlActions,
)(VersionControlView)

export default ConnectedGitComponent
