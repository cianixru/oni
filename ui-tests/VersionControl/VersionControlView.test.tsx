import { shallow, mount } from "enzyme"
import { shallowToJson } from "enzyme-to-json"
import * as React from "react"

import { SectionTitle, Title } from "./../../browser/src/UI/components/SectionTitle"

import {
    DefaultState,
    VersionControlState,
} from "./../../browser/src/Services/VersionControl/VersionControlStore"
import { VersionControlView } from "./../../browser/src/Services/VersionControl/VersionControlView"
import CommitMessage, {
    Explainer,
} from "./../../browser/src/UI/components/VersionControl/CommitMessage"
import Commits from "./../../browser/src/UI/components/VersionControl/Commits"
import Help from "./../../browser/src/UI/components/VersionControl/Help"
import Staged, { LoadingHandler } from "./../../browser/src/UI/components/VersionControl/Staged"
import VersionControlStatus from "./../../browser/src/UI/components/VersionControl/Status"

const noop = jest.fn()

jest.mock("./../../browser/src/neovim/SharedNeovimInstance", () => ({
    getInstance: () => ({
        bindToMenu: () => ({
            setItems: jest.fn(),
            onCursorMoved: {
                subscribe: jest.fn(),
            },
        }),
    }),
}))

const IDs: any = {
    modified: "modified",
    commits: "commits",
    untracked: "untracked",
    staged: "staged",
    commitAll: "commit_all",
}

const makePromise = (arg?: any) => Promise.resolve(arg)

jest.mock("../../browser/src/UI/components/Sneakable", () => {
    const React = require("react") // tslint:disable-line
    return { Sneakable: () => <div /> }
})

describe("<VersionControlView />", () => {
    const state = { ...DefaultState, activated: true, hasFocus: true }
    const container = shallow(
        <VersionControlView
            {...state}
            commit={noop}
            filesToCommit={["/test.txt"]}
            setLoading={noop}
            IDs={IDs}
            showHelp={true}
            committing={false}
            cancelCommit={noop}
            updateCommitMessage={noop}
            commits={[]}
            message={[]}
            selectedItem={null}
            loadingSection={null}
            loading={false}
            getStatus={() => makePromise({})}
        />,
    )
    it("renders without crashing", () => {
        expect(container.length).toBe(1)
    })

    it("should render an untracked, staged and modified section", () => {
        const container = mount(
            <VersionControlView
                {...state}
                commit={noop}
                setLoading={noop}
                filesToCommit={["/test.txt"]}
                IDs={IDs}
                loading={false}
                loadingSection={null}
                showHelp={false}
                committing={false}
                cancelCommit={noop}
                updateCommitMessage={noop}
                commits={[]}
                message={[]}
                selectedItem={null}
                getStatus={() => makePromise({})}
            />,
        )
        const staged = container.find(Staged)
        expect(staged.length).toBe(1)
        const untrackedAndModified = container.find(VersionControlStatus)
        expect(untrackedAndModified.length).toBe(2)
        const commits = container.find(Commits)
        expect(commits.length).toBe(1)
    })

    it("shouldn't show a section if it has no content", () => {
        const wrapper = shallow(
            <VersionControlStatus
                onClick={noop}
                toggleVisibility={noop}
                visibility={true}
                titleId="modified"
                selectedId="file1"
                icon="M"
                files={null}
            />,
        )
        expect(wrapper.find(SectionTitle).length).toBe(0)
    })

    it("should match the last recorded snapshot unless a change was made", () => {
        const wrapper = shallow(
            <VersionControlStatus
                titleId="modified"
                visibility={true}
                toggleVisibility={noop}
                onClick={noop}
                selectedId="file1"
                icon="M"
                files={["test1", "test2"]}
            />,
        )
        expect(shallowToJson(wrapper)).toMatchSnapshot()
    })

    it("Should render only the title if not visible", () => {
        const wrapper = mount(
            <Staged
                files={[]}
                filesToCommit={["test.txt"]}
                titleId="staged"
                visible={false}
                handleCommitOne={jest.fn()}
                handleCommitMessage={jest.fn()}
                handleCommitCancel={jest.fn()}
                handleCommitAll={jest.fn()}
                toggleVisibility={noop}
                selectedId="commit_all"
                icon="cross"
                loading={false}
                handleSelection={noop}
            />,
        )

        expect(wrapper.find(SectionTitle).length).toBe(1)
        expect(wrapper.children.length).toBe(1)
    })

    it("should render a loading spinner if loading is true", () => {
        const wrapper = mount(
            <Staged
                files={["test.txt"]}
                filesToCommit={["test.txt"]}
                selectedToCommit={() => false}
                titleId="staged"
                visible
                handleCommitOne={jest.fn()}
                handleCommitMessage={jest.fn()}
                handleCommitCancel={jest.fn()}
                handleCommitAll={jest.fn()}
                toggleVisibility={noop}
                selectedId="commit_all"
                icon="cross"
                loading
                handleSelection={noop}
            />,
        )
        expect(wrapper.find(LoadingHandler).length).toBe(2)
    })

    it("should render an Input in place of the commit all explainer", () => {
        const wrapper = mount(
            <Staged
                visible
                loading
                icon="cross"
                titleId="staged"
                files={["test.txt"]}
                filesToCommit={[]}
                selectedToCommit={() => true}
                handleCommitOne={jest.fn()}
                handleCommitMessage={jest.fn()}
                handleCommitCancel={jest.fn()}
                handleCommitAll={jest.fn()}
                toggleVisibility={noop}
                selectedId="commit_all"
                handleSelection={noop}
            />,
        )
        expect(wrapper.find(CommitMessage).length).toBe(1)
    })

    it("should render help if show help is true", () => {
        expect(container.find(Help).length).toBe(1)
    })

    it("should render help prompt if show help is false", () => {
        container.setProps({ showHelp: false })
        expect(
            container
                .find(Title)
                .dive()
                .text(),
        ).toBe('To show help press "?"')
    })
})
