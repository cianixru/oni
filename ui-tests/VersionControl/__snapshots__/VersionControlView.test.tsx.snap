// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VersionControlView /> should match the last recorded snapshot unless a change was made 1`] = `
<div>
  <SidebarSectionTitle
    active={true}
    count={2}
    isSelected={false}
    onClick={[MockFunction]}
    testId="modified-2"
    title="modified"
  />
  <File
    file="test1"
    icon="M"
    isSelected={false}
    key="test1"
    onClick={[MockFunction]}
  />
  <File
    file="test2"
    icon="M"
    isSelected={false}
    key="test2"
    onClick={[MockFunction]}
  />
</div>
`;
